### PR TITLE
mimxrt: Add the hw i2c class to machine.

### PR DESCRIPTION
**Pull request by robert-hh**
Advantage over SoftI2C:
- Full protocol support.
- Higher data rate -  up to ~3 MHZ.

### CODE REVIEW

File: ports/mimxrt/machine_i2c.c [hunk 1]
```diff
@@ -50,10 +50,9 @@ typedef struct _machine_i2c_obj_t {
     LPI2C_Type *i2c_inst;
     uint8_t i2c_id;
     uint8_t i2c_hw_id;
-    bool transfer_busy;
-    bool transfer_error;
-    bool nack_flag;
     uint8_t mode;
+    bool transfer_busy;
+    status_t transfer_status;
```
Comment (kadamski) [inline ports/mimxrt/machine_i2c.c:53]: Many of the changes in this commit are not related to the force stop mode and it would be better to just move them to the previous commit, which would make less confusion.

File: ports/mimxrt/machine_i2c.c [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+    self->transfer_busy = false;
+    if (status == kStatus_LPI2C_Nak) {
+        self->nack_flag = true;
+    } else {
+        self->transfer_error = (status != kStatus_Success);
+    }
+}
+
+    lpi2c_master_handle_t g_master_handle;
```
[the remark this answers is not in the document]
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c]: That line 165 does not exist in my file.

File: ports/mimxrt/machine_i2c.c [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    int i2c_id = mp_obj_get_int(args[ARG_id].u_obj);
+    if (i2c_id < 0 || i2c_id >= MICROPY_HW_I2C_NUM) {
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("I2C(%d) doesn't exist"), i2c_id);
+    }
+
+    // Get I2C Object.
+    machine_i2c_obj_t *self = m_new_obj(machine_i2c_obj_t);
+    self->base.type = &machine_hw_i2c_type;
+    self->i2c_id = i2c_id;
+    self->i2c_hw_id = i2c_index_table[i2c_id];  // the hw spi number 1..n
```
[the remark this answers is not in the document]
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:121]: Indeed a copy/paste glitch from the template file in the comment.

File: ports/mimxrt/machine_i2c.c [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+static void lpi2c_master_callback(LPI2C_Type *base, lpi2c_master_handle_t *handle, status_t status, void *self_in) {
+    machine_i2c_obj_t *self = (machine_i2c_obj_t *)self_in;
+
+    self->transfer_busy = false;
+    if (status == kStatus_LPI2C_Nak) {
+        self->nack_flag = true;
+    } else {
+        self->transfer_error = (status != kStatus_Success);
```
[the remark this answers is not in the document]
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c]: You are working on an old version of the file. So some of your comments do not apply

File: ports/mimxrt/machine_i2c.c [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("I2C(%d) doesn't exist"), i2c_id);
+    }
+
+    // Get I2C Object.
+    machine_i2c_obj_t *self = m_new_obj(machine_i2c_obj_t);
+    self->base.type = &machine_hw_i2c_type;
+    self->i2c_id = i2c_id;
+    self->i2c_hw_id = i2c_index_table[i2c_id];  // the hw spi number 1..n
+    self->i2c_inst = i2c_base_ptr_table[self->i2c_hw_id];
+    self->mode = 0;
```
[the remark this answers is not in the document]
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:123]: self->mode is not used. Was planned for master/slave mode.

File: ports/mimxrt/boards/MIMXRT1020_EVK/mpconfigboard.h [hunk 6]
```diff
@@ -13,7 +13,8 @@
 #define MICROPY_HW_NUM_PIN_IRQS (3 * 32)
 
 // Define mapping logical UART # to hardware UART #
-// D3/D5   LPUART1  Not usable, Since D3 is blocked.
+// RX/TX   HW-UART    Logical I2C
+// D3/D5   LPUART1    Not usable, Since D3 is blocked.
```
[the remark this answers is not in the document]
Comment (robert-hh) [reply ports/mimxrt/boards/MIMXRT1020_EVK/mpconfigboard.h:17]: No. It is related to UART. But it is a part of the mpconfigport.h file.

File: ports/mimxrt/machine_i2c.c (hunk 1 above)
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:53]: This part of the code is changed in further commits. So I leave it as it is. And I will definitely not start to split up commits, which will be squashed at the end anyhow. Please make you comment to the final state of the files, not some intermediate states. That is pointless.

Comment (kadamski) [reply ports/mimxrt/machine_i2c.c:52]: But how could I know if you intend to squash commits and which of them you intend to squash?

Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:52]: Damien will probably squash all commits of this PR into a single one, when he merges the commit. So do not look at individual commits here, because the only purpose of the PR is providing HW I2C.

Comment (kadamski) [reply ports/mimxrt/machine_i2c.c:52]: Ach, ok. I find this strategy strange as, well, what should be the commit message of such a squashed commit? :) But, well, its up to Damien. I'll be reviewing the result only from now on.

**Comment by robert-hh** [comment]
@dpgeorge This PR is ready to go.

File: ports/mimxrt/machine_i2c.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    } else if (self->transfer_status == kStatus_LPI2C_Nak) {
+        return -MP_ENODEV;
+    } else {
+        return -MP_EIO;
+    }
+}
+
+STATIC const mp_machine_i2c_p_t machine_i2c_p = {
+    .transfer = mp_machine_i2c_transfer_adaptor,
+    .transfer_single = machine_i2c_transfer_single,
```
Comment (dpgeorge) [inline ports/mimxrt/machine_i2c.c:207]: I guess you needed to use the adaptor because there was no easy way to transfer multiple segments?  It's not a problem, just wondering.

File: ports/mimxrt/machine_i2c.c [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    ret = LPI2C_MasterTransferNonBlocking(self->i2c_inst, &g_master_handle, &masterXfer);
+    if (ret != kStatus_Success) {
+        return -MP_EIO;
+    }
+    //  Wait for the transfer to complete
+    while (self->transfer_busy) {
+        MICROPY_EVENT_POLL_HOOK
+    }
+
+    // Transfer will not send a stop in case of errors like NAK. So it#s done here.
```
Comment (dpgeorge) [inline ports/mimxrt/machine_i2c.c:191]: `it's` instead of `it#s`

File: ports/mimxrt/machine_i2c.c [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        return -MP_EIO;
+    }
+}
+
+STATIC const mp_machine_i2c_p_t machine_i2c_p = {
+    .transfer = mp_machine_i2c_transfer_adaptor,
+    .transfer_single = machine_i2c_transfer_single,
+};
+
+const mp_obj_type_t machine_hw_i2c_type = {
```
Comment (dpgeorge) [inline ports/mimxrt/machine_i2c.c:210]: maybe call it `machine_i2c_type` to match `machine_spi_type`?

File: ports/mimxrt/machine_i2c.c [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    mp_printf(print, "I2C(%u, freq=%u)",
+        self->i2c_id, self->master_config->baudRate_Hz);
+}
+
+mp_obj_t machine_i2c_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_id, ARG_freq, ARG_drive};
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_id, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_freq, MP_ARG_INT, {.u_int = DEFAULT_I2C_FREQ} },
+        { MP_QSTR_drive, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = DEFAULT_I2C_DRIVE} },
```
Comment (dpgeorge) [inline ports/mimxrt/machine_i2c.c:101]: what arguments can be passed to `drive`?

File: ports/mimxrt/machine_i2c.c (hunk 7 above)
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:207]: It was just a copy of method wihtout much thinking.

File: ports/mimxrt/machine_i2c.c (hunk 8 above)
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:191]: OK. Will be changed

File: ports/mimxrt/machine_i2c.c (hunk 10 above)
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:101]: The same as for the Pin module and and the SPI module. Values 0 - 6 or Pin.POWER_0 to Pin.POWER_6. It changes the hardware pin impedance.  Since pins are assigned fixed to the I2C id's, not Pin settings can be made. But obviously this is only critical for the low edge. And since the I2C speed is low, it is less important than for SPI.

File: ports/mimxrt/machine_i2c.c (hunk 9 above)
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:210]: I can change that without conflicts.

File: ports/mimxrt/machine_i2c.c (hunk 7 above)
Comment (robert-hh) [reply ports/mimxrt/machine_i2c.c:207]: But in fact the NXP API performs all the frame handling like handling start and stop conditions,  and expect a single message to be sent.

**Comment by robert-hh** [comment]
The PR is updated with the suggested changes.

**Comment by dpgeorge** [comment]
Thank you!  Squashed and merged in 98c5703027e7db9a030f747f4dd1684407230656

**Comment by robert-hh** [comment]
Thanks a lot. The next PR is most likely Ethernet support for the MIMXRT boards and Teensy 4.1. As far as I can tell it is already working fine. I just keep it on the bench for a while for testing and further thoughts.
